VisualPinball.Unity/VisualPinball.Unity/VPT/Mech/ScoreMotorComponent.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Visual Pinball Engine
+// Copyright (C) 2021 freezy and VPE Team
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// ReSharper disable InconsistentNaming
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using VisualPinball.Engine.Game.Engines;
+
+namespace VisualPinball.Unity
+{
+	public delegate void ScoreMotorResetCallback(float score);
+
+	public readonly struct ScoreMotorResetEventArgs
+	{
+		public readonly string Id;
+		public readonly float Score;
+		public readonly ScoreMotorResetCallback Callback;
+
+		public ScoreMotorResetEventArgs(string id, float score, ScoreMotorResetCallback callback)
+		{
+			Id = id;
+			Score = score;
+			Callback = callback;
+		}
+	}
+
+	public delegate void ScoreMotorAddPointsCallback(float points);
+
+	public readonly struct ScoreMotorAddPointsEventArgs
+	{
+		public readonly string Id;
+		public readonly float Points;
+		public readonly ScoreMotorAddPointsCallback Callback;
+
+		public ScoreMotorAddPointsEventArgs(string id, float points, ScoreMotorAddPointsCallback callback)
+		{
+			Id = id;
+			Points = points;
+			Callback = callback;
+		}
+	}
+
+	public enum ScoreMotorMode
+	{
+		Reset = 0,
+		AddPoints = 1
+	}
+
+	public enum ScoreMotorAction
+	{
+		Wait = 0,
+		Increase = 1
+	}
+
+	[Serializable]
+	public class ScoreMotorActions
+	{
+		public List<ScoreMotorAction> Actions = new List<ScoreMotorAction>();
+	}
+
+	[AddComponentMenu("Visual Pinball/Game Item/Score Motor")]
+	public class ScoreMotorComponent : MonoBehaviour, ISwitchDeviceComponent
+	{
+		public const int MaxIncrease = 5;
+
+		[Unit("\u00B0")]
+		[Tooltip("The total number of degrees in one turn.")]
+		public int Degrees = 120;
+
+		[Unit("ms")]
+		[Tooltip("Amount of time, in milliseconds to move one turn.")]
+		public int Duration = 769;
+
+		[Tooltip("The total number of steps per turn.")]
+		[Min(MaxIncrease)]
+		public int Steps = 6;
+
+		[Tooltip("Disable to allow single point scores while score motor running.")]
+		public bool BlockScoring = true;
+
+		public List<ScoreMotorActions> ScoreMotorActionsList = new List<ScoreMotorActions>()
+		{
+			new ScoreMotorActions(),
+			new ScoreMotorActions(),
+			new ScoreMotorActions(),
+			new ScoreMotorActions(),
+			new ScoreMotorActions()

```

Comment:
Shouldn't there be 6 of them, given `Steps` is 6 per default?

Reply:
I couldn't figure out how to do multi-dimensional arrays with `SerializedProperty`, so I had to make a list of a `ScoreMotorActions` classes that had a list of `ScoreMotorAction` enums.

To make this a little clearer, I renamed the class to `ScoreMotorTiming`

So we will always have 5 timings, with each one having "step" amount of actions.

It would be nice if I could have done this with something like:

`public List<ScoreMotorTiming> ScoreMotorTimingList = Enumerable.Repeat(new ScoreMotorTiming(), ScoreMotorComponent.MaxIncrease).ToList();`

But then the timing list is reset anytime the inspector is open.